packages/crud/src/vaadin-crud-grid.js

```diff
@@ -164,42 +194,64 @@ class CrudGrid extends IncludedMixin(Grid) {
 
           filter.appendChild(textField);
           root.appendChild(filter);
-        }
-
-        if (this.noSort && this.noFilter) {
+        } else if (this.noSort && this.noFilter) {
+          // Neither sorter nor filter are enabled, just add the label
           root.textContent = label;
         }
       };
     }
-
-    parent.appendChild(col);
-    return col;
   }
 
-  /** @private */
-  __createColumns(parent, object, path) {
-    if (typeof object === 'object') {
+  /**
+   * Creates the column structure for the (sub)object.
+   *
+   * @param {HTMLElement} parent May be the crud-grid or a column group.
+   * @param {Object} object The object to create the sub-columns for.
+   * @param {string} path The property path from the root item to the object.
+   * @param {number} depth The depth of the object in the object hierarchy.
+   * @private
+   **/
+  __createColumns(parent, object, path, depth) {
+    if (object && typeof object === 'object') {
+      // Iterate over the object properties
       Object.keys(object).forEach((prop) => {
         if (!this.include && this.exclude && this.exclude.test(prop)) {
           return;
         }
-        const newPath = (path ? `${path}.` : '') + prop;
-        if (object[prop] && typeof object[prop] === 'object') {
-          const group = this.noHead ? parent : this.__createGroup(parent, newPath, object[prop]);
-          this.__createColumns(group, object[prop], newPath);
-        } else {
-          this.__createColumn(parent, newPath);
+        // Sub-object of the current object
+        const subObject = object[prop];
+        // Full path to the sub-object
+        const subObjectPath = path ? `${path}.${prop}` : prop;
+
+        // The column element for the sub-object
+        let subObjectColumn = parent;
+        if (!this.noHead && depth > 1) {
+          const isSubObject = subObject && typeof subObject === 'object';
+          // If the sub-object is an actual object, create a column group with the property
+          // name as the header text, otherwise create a group without a header
+          subObjectColumn = this.__createGroup(parent, isSubObject ? prop : undefined);
         }
+
+        // Run recursively for the sub-object level
+        this.__createColumns(subObjectColumn, subObject, subObjectPath, depth - 1);
       });
+    } else if (depth > 1) {
+      // The object has been fully traversed, but empty wrapping column
+      // groups are still needed to complete the full object depth
+      const group = this.noHead ? parent : this.__createGroup(parent);
+      this.__createColumns(group, undefined, path, depth - 1);
```

Comment:
I tried to remove this branch and the tests didn't fail. The CRUD dev page also appeared not to be affected. That being so, is this still relevant?

Reply:
Meanwhile, the branch is quite suspicious as a whole. What is the scenario in which `depth > 1 && typeof object !== 'object'`?

Reply:
It's relevant, with the following item structure:
```js
{
  firstName: 'John',
  foo: {
    bar: {
      baz: '1',
      qux: '2'
    }
  }
},
```
you'd end up with:
![Screenshot 2021-11-23 at 9 44 25](https://user-images.githubusercontent.com/1222264/142986760-cbcd82ec-92d0-4348-bdaa-2158dd7242d4.png)

Will add the missing test

Reply:
Oh, now I get it. 👍  Yeah, unit tests would be necessary then.